### PR TITLE
Enable Pipeline Parallelism on Jax runner

### PR DESCRIPTION
**Pull request by Chenyaaang**
# Description

The implementation for Pipeline Parallelism are splitted into the following small PRs.

- [ ] Util functions for PP on Jax (https://github.com/vllm-project/tpu-inference/pull/1042)
- [ ] worker changes (https://github.com/vllm-project/tpu-inference/pull/1043)
- [x] runner changes (The current PR)
- [ ] platform changes
- [ ] torchax changes
- [ ] Jax changes
- [ ] Single host changes (this will be in vLLM)
- [ ] Multi host changes (Ray)

This PR is to modify Jax runner to support PP.
- `tpu_jax_runner.py` pass in `intermediate_tensors` to `_execute_model`
- `compilation_manager.py` add PP to the precompile backbone
-  `persistent_batch_manager` update the internal status for all non-last rank, based on the scheduler output's last selected tokens.

# Tests

 E2E test has verified the whole PP implementation works properly.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/runner/persistent_batch_manager.py [hunk 1]
```diff
@@ -173,15 +174,42 @@ def update_states(self, scheduler_output: "VllmSchedulerOutput",
...
         for i, req_id in enumerate(req_data.req_ids):
             req_state = self.requests[req_id]
             num_computed_tokens = req_data.num_computed_tokens[i]
             new_block_ids = req_data.new_block_ids[i]
             resumed_from_preemption = req_data.resumed_from_preemption[i]
+            num_output_tokens = req_data.num_output_tokens[i]
 
             # Update the cached states.
             req_state.num_computed_tokens = num_computed_tokens
+            req_index = self.input_batch.req_id_to_index.get(req_id)
```
Comment (py4) [inline tpu_inference/runner/persistent_batch_manager.py:187]: Can you please add unit tests for changes in persistent batch manager? 
and also e2e test for the feature? 
Thanks!

Comment (kyuyeunk) [reply tpu_inference/runner/persistent_batch_manager.py:187]: +1 on adding tests

File: tpu_inference/runner/persistent_batch_manager.py [hunk 2]
```diff
@@ -173,15 +174,42 @@ def update_states(self, scheduler_output: "VllmSchedulerOutput",
                     )
 
         # Update the states of the running/resumed requests.
+        is_last_rank = get_pp_group().is_last_rank
```
Comment (yaochengji) [inline tpu_inference/runner/persistent_batch_manager.py]: When not using pp, is `is_last_rank` always True?

Comment (Chenyaaang) [reply tpu_inference/runner/persistent_batch_manager.py]: Yes, if pp=1, `is_last_rank` and `is_first_rank` are both True.

File: tpu_inference/runner/persistent_batch_manager.py (hunk 1 above)
Comment (Chenyaaang) [reply tpu_inference/runner/persistent_batch_manager.py:187]: Added.

**Comment by Chenyaaang** [comment]
The CI is green except for the "kv_cache_utils" problems reported in the group chat and also broke in the main branch. https://buildkite.com/tpu-commons/tpu-inference-ci/builds/5940

File: tpu_inference/runner/persistent_batch_manager.py (hunk 1 above)
Comment (py4) [reply tpu_inference/runner/persistent_batch_manager.py:187]: Thanks for adding the test. Can you please update https://github.com/vllm-project/tpu-inference/blob/main/.buildkite/parallelism/PP.yml so that CI runs your added e2e test? By default tests at e2e are not run unless they are configured in a yaml 

for reference you can look at https://github.com/vllm-project/tpu-inference/blob/main/.buildkite/features/Speculative_Decoding-_Ngram.yml

CC: @jcyang43

File: tests/e2e/test_pipeline_parallel.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    test_prompts: list,
+    sampling_params: SamplingParams,
+):
+    """
+    Test pipline parallelism works on Jax models
+
+    Equivalent to:
+    python examples/offline_inference.py --tensor_parallel_size=1 --pipeline_parallel_size=2
+    """
+    # Test with data parallelism enabled
```
Comment (yixinshi) [inline tests/e2e/test_pipeline_parallel.py]: data -> pipeline?

File: tests/e2e/test_pipeline_parallel.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    MODEL_IMPL_TYPE=vllm python examples/offline_inference.py --tensor_parallel_size=2 --pipeline_parallel_size=2
+    """
+
+    os.environ['MODEL_IMPL_TYPE'] = 'vllm'
+    # Test with data parallelism enabled
+    outputs = _run_inference_with_config(
+        model_name=model_name,
+        test_prompts=test_prompts,
+        sampling_params=sampling_params,
+        tensor_parallel_size=1,
```
Comment (yixinshi) [inline tests/e2e/test_pipeline_parallel.py:136]: This doesn't match line 127

File: tpu_inference/runner/compilation_manager.py [hunk 5]
```diff
@@ -266,10 +285,28 @@ def _precompile_backbone_text_only(self) -> None:
                                                   dp_sharding)
             positions = self._create_dummy_tensor((num_tokens, ), jnp.int32,
                                                   dp_sharding)
-            self._precompile_backbone_helper("backbone",
-                                             input_ids=input_ids,
-                                             positions=positions,
-                                             inputs_embeds=None)
+            is_first_rank = self.runner.is_first_rank
+            is_last_rank = self.runner.is_last_rank
+            if not is_first_rank:
```
Comment (yixinshi) [inline tpu_inference/runner/compilation_manager.py]: nit: positive conditions normally go first in "if" statement.

File: tests/e2e/test_pipeline_parallel.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    try:
+        outputs = llm.generate(test_prompts, sampling_params)
+        return outputs
+    finally:
+        del llm
+        # Wait for TPUs to be released
+        time.sleep(5)
+
+
+@pytest.mark.skip(reason="PP is not fully enabled.")
```
Comment (samos123) [inline tests/e2e/test_pipeline_parallel.py:82]: not sure if intentional, but these tests are not being run at all since the tests are being skipped.

Comment (Chenyaaang) [reply tests/e2e/test_pipeline_parallel.py:82]: Yes, I intentionally set to skip since PP isn't fully enabled.

Comment (samos123) [reply tests/e2e/test_pipeline_parallel.py:82]: This is causing the supportability matrix to get updated to show that it's fully supported. See here: https://docs.vllm.ai/projects/tpu/en/latest/recommended_models_features/#parallelism-support

People incorrectly assumed that pipeline parallelism is fully tested and enabled.
